hyperspy/learn/mva.py

```diff
@@ -1636,7 +1642,747 @@ def undo_treatments(self):
+                center = np.dot(a, b).sum(axis=0)
+                cluster_centers[i, :] = cluster_centers[i, :] + center
+                cluster_centers[i, :] = cluster_centers[i, :] 
+            else:
+                cluster_centers[i, :] = clusterdata[clus_index].sum(axis=0)
+
+        # this sorts the labels based on clustersize for high to low
+        # i.e. point with laster number of points first
+        idx = np.argsort(clustersizes)[::-1]
+        lut = np.zeros_like(idx)
+        lut[idx] = np.arange(n_clusters)
+        sorted_labels = lut[labels]
+        shape = (n_clusters, self.data.shape[0])
+        cluster_labels = np.full(shape, np.nan)
+        sorted_cluster_centers = np.zeros_like(cluster_centers)            
+        # now create the labels from these sorted labels
+        for i in range(n_clusters):
+            cluster_labels[i, navigation_mask] = \
+                np.where(sorted_labels == i, 1, 0)
+            sorted_cluster_centers[i,:] = cluster_centers[lut[i],:]\
+                /clustersizes[lut[i]]                
+
+
+
+        return sorted_labels,cluster_labels, sorted_cluster_centers
+
+
+    def cluster_analysis(self,
+                         n_clusters=None,
+                         scaling="norm",
+                         scaling_kwargs={},
+                         use_decomposition_results=True,
+                         use_decomposition_results_for_centers=False,
+                         number_of_components=None,
+                         navigation_mask=None,
+                         signal_mask=None,
+                         algorithm='kmeans',
+                         return_info=False,
+                         **kwargs):
+        """
+        Cluster analysis of a signal or decomposition results of a signal
+
+        Results are stored in `learning_results`.
+
+        Parameters
+        ----------
+        n_clusters : int
+            Number of clusters to find.
+        scaling : {"standard","norm","minmax",None or scikit learn scaling method}
+            default: 'norm'
+            Preprocessing the data before cluster analysis requires scaling
+            the data to be clustered to similar scales. Standard scaling
+            adjusts each feature to have uniform variation. Norm scaling
+            adjusts treats the set of features like a vector and
+            each measurement is scaled to length 1.
+            You can also pass one of the scikit-learn preprocessing
+            scale_method = import sklearn.processing.StandadScaler
+            scaling = scale_method
+            See scaling methods in scikit-learn preprocessing for further
+            details.
+        scaling_kwargs : dict
+            Additional parameters passed to the cluster scaling algorithm.
+            See sklearn.preprocessing scaling methods for further details
+        use_decomposition_results : {True,False,"decomposition","bss"}
```

Comment:
I think that it would be more flexible and clear replacing this keyword argument by `X` (like in sklearn). It could take "bss", "decomposition", "signal" or any signal with the right dimensions. It could be left without default value to force the user to think about what they are doing.

Reply:
Change made - haven't used X but user has to supply "bss","decomposition" or "signal"